packages/lesswrong/server/emails/renderEmail.tsx

```diff
@@ -240,6 +240,12 @@ export const wrapAndSendEmail = async ({user, to, from, subject, body}: {
   if (!to && !user) throw new Error("No destination email address for logged-out user email");
   const destinationAddress = to || getUserEmail(user)
   if (!destinationAddress) throw new Error("No destination email address for user email");
+  const _reasonUserCantReceiveEmails = user && reasonUserCantReceiveEmails(user)
+  if(user && !!_reasonUserCantReceiveEmails) {
```

Comment:
It seems like the main crux of this PR is "will the addition of the check to many sources of email-sends cause issues." I could try to figure this out, but could I start from the reasoning y'all used? Or, if you didn't think through it carefully, could you do so and write up your results?

Reply:
We thought it through reasonably carefully (Sarah and I did this in a pairing session), and I think this is about as good as we can do to solve the narrow problem (that checking "unsubscribe from all" hardly blocks any emails).

The crux of the decision of how to do this was whether to:
1. Make `unsubscribeFromAll` independent of all the other notification settings (so e.g. you can set both "receive DM notifications by email" and "unsubscribe from all", and this will mean you DO NOT receive DM emails, but if you uncheck "unsubscribe from all" you will start receiving them again). The benefit of this is that you don't lose all your settings state if you check "unsubscribe from all", but the downside is that the individual notification settings no longer define the behaviour on their own, so you could easily accidentally set "unsubscribe from all" and forget about it and then be confused as to why you're not receiving emails when you turn them on individually
2. Make `unsubscribeFromAll` effectively control all the other settings, so setting it turns off all the other settings individually. This has the downside that you lose all your state once you check "unsubscribe from all", but the benefit that all the individual settings completely define the behaviour

We went with (1), except for the digest subscription (for which we went with (2)). We did it this way essentially because there can be a digest ad on the frontpage with a button to subscribe, which makes the problem described in (1) much more likely ("you could easily accidentally set "unsubscribe from all" and forget about it and then be confused as to why you're not receiving emails when you turn them on individually").

I think the behaviour in this PR is strictly better than the current situation (where the button basically doesn't work at all), but is still pretty suboptimal and the correct solution is "(1) + a lot of UX changes to make it more obvious what is happening"

Reply:
I'm on board with that hacky solution to the UI.

But I'm a bit worried that there's some usage of `wrapAndSendEmail` that assumed that it _wouldn't_ check all the various reasons why a user shouldn't get an email. For example, maybe the LW verification email uses `wrapAndSendEmail`, and then there'd be a catch 22. You haven't persuaded me that you've thought about that sort of thing.

Reply:
Wow good catch, that was indeed the case for that example. I've fixed that one and checked all the other uses (of which there are only about 10), and I don't think there are any others like this

Reply:
Nice! I wondered when I gave that example if it would in fact be true but the only one. That's the only comment I had, you're good now.